pallets/system/src/data_root_builder.rs

```diff
@@ -0,0 +1,355 @@
+			hex!("3630646564316635616236373261373132376261").into(),
+			hex!("3262313166316464333935353666623261623432").into(),
+		];
+
+		let leaves = avail_data
+			.iter()
+			.map(|xt| Sha256::hash(&xt))
+			.collect::<Vec<[u8; 32]>>();
+
+		let data_tree = MerkleTree::<Sha256>::from_leaves(&leaves);
+		let proof = data_tree.proof(&[1usize]);
+		let root_proof = proof.proof_hashes().to_vec();
+		assert_eq!(root_proof, vec![
+			hex!("754B9412E0ED7907BDF4B7CA5D2A22F5E129A03DEB1F4E1C1FE42D322FDEE90E"),
+			hex!("8D6E30E494D17D7675A94C3C614467FF8CCE35201C1056751A6E9A100515DAF9")
+		]);
+	}
+
+	#[test]
+	fn test_single_merkle_proof() {
+		let empty_vec: Vec<[u8; 32]> = vec![];
+
+		let avail_data: Vec<Vec<u8>> =
+			vec![hex!("3435346666383063303838616137666162396531").to_vec()];
+
+		let leaves = avail_data
+			.iter()
+			.map(|xt| Sha256::hash(&xt))
+			.collect::<Vec<[u8; 32]>>();
+
+		let data_tree = MerkleTree::<Sha256>::from_leaves(&leaves);
+		let proof = data_tree.proof(&[0usize]);
+		let root_proof = proof.proof_hashes().to_vec();
+		// here the proof is shown empty because the root itself is the proof as there is only one appdata extrinsic
+		assert_eq!(root_proof, empty_vec);
+	}
+
+	///using rs-merkle proof verify function
+	#[test]
+	fn verify_merkle_proof() {
+		let avail_data: Vec<Vec<u8>> = vec![
+			hex!("3033333166613733656565636362653465323235").into(),
+			hex!("3630646564316635616236373261373132376261").into(),
+			hex!("3262313166316464333935353666623261623432").into(),
+			hex!("6433326630643762346634306264346563323665").into(),
+		];
+		let leaves = avail_data
+			.iter()
+			.map(|xt| Sha256::hash(&xt))
+			.collect::<Vec<[u8; 32]>>();
+
+		let merkle_tree = MerkleTree::<Sha256>::from_leaves(&leaves);
+		let indices_to_prove = vec![3];
+		let leaves_to_prove = leaves.get(3..4).ok_or("can't get leaves to prove").unwrap();
+
+		let proof = merkle_tree.proof(&indices_to_prove);
+		let root = merkle_tree
+			.root()
+			.ok_or("couldn't get the merkle root")
+			.unwrap();
+
+		assert!(proof.verify(root, &indices_to_prove, leaves_to_prove, leaves.len()));
+	}
+
+	#[test]
+	fn verify_nodata_merkle_proof() {
+		let avail_data: Vec<Vec<u8>> = vec![];
+
+		let leaves = avail_data
+			.iter()
+			.map(|xt| Sha256::hash(&xt))
+			.collect::<Vec<[u8; 32]>>();
+		let leaves_to_prove = if let Ok(leaves) = leaves.get(0).ok_or("can't get leaves to prove") {
+			leaves
+		} else {
+			&[0u8; 32]
+		};
+		assert_eq!(leaves_to_prove, &[0u8; 32]);
+	}
+
+	fn encoded_submit_call<A: Into<AppId>>(app_id: A) -> AppExtrinsic {
+		let data = hex!("5D0284001CBD2D43530A44705AD088AF313E18F80B53EF16B36177CD4B77B846F2A5F07C01C44755794EA949E9410390CB4CE07FE2D8068656185B5AB9B43EEF934C3680478968C1F83E360A5D942FE75E9D58E49106A8E8B23601CBC6A633D80E5D089D83A4000400030000001D01A46868616A6B616E636B61206C61682069616B6A206361697568206162206169616A6820612067616861").to_vec();
+		AppExtrinsic {
+			app_id: app_id.into(),
+			data,
+		}
+	}
+
```

Comment:
do we have test cases checking root parity between the two libs? also, these tests look to be mainly focussing on correct proof of inclusion. do we have checks about correctness of data root generation?

Reply:
Yes, it is in `runtime/src/data_root_tests.rs:67`. The check generates the same expected output for both libraries. In order to verify `rs-merkle`, you have to enable the feature like:
```Bash
$> cargo test --workspace --features "frame-system/force-rs-merkle"
```

**Note** that any  Data Root test has been moved to `runtime` level because there is a dependency between the `Call::DataAvailility(da_ctr::Call::submit_data{...})` and the implementation of `DataRootFilter`. That will **make code upgradable** and remove lot of duplicate/hardcode deserialization code.